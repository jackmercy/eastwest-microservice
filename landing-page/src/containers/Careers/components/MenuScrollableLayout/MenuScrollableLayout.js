import Link from 'next/link';
import { useRouter } from 'next/router';
import { memo, useCallback, useEffect, useMemo, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { scrollToElement } from '../../../../utils';
import { useEventListener } from '../../../../hooks';
import BREAKPOINTS from '../../../../constants/BREAKPOINTS';

import useStyles from './menuScrollableLayout.styles';

const MenuScrollableLayout = ({
  menuConfigs,
  children,
  className,
  menuClassName,
  contentClassName,
}) => {
  const classes = useStyles();
  const router = useRouter();
  const ids = useMemo(() => menuConfigs.map((item) => item.id), [menuConfigs]);
  const savingIdRects = useRef([
    // eslint-disable-next-line prefer-spread
    ...Array.apply(null, Array(ids.length)).map(() => ({
      offset: 0,
      limitUp: 0,
      limitDown: 0,
    })),
  ]);
  const [activatedMenu, setActiveMenu] = useState(ids[0]);

  useEffect(() => {
    setTimeout(() => {
      if (window.innerWidth > BREAKPOINTS.maxMd) {
        ids.forEach((id, index) => {
          const element = document.getElementById(id);
          if (element) {
            const offset = element.offsetTop;
            savingIdRects.current[index].offset = offset;

            if (index === 0) {
              savingIdRects.current[index].limitUp = 0;
            } else {
              const limit =
                savingIdRects.current[index - 1].offset +
                (offset - savingIdRects.current[index - 1].offset) / 2;
              savingIdRects.current[index - 1].limitDown = limit;
              savingIdRects.current[index].limitUp = limit;
            }
            if (index === ids.length - 1) {
              const { body, documentElement: html } = document;
              const height = Math.max(
                body.scrollHeight,
                body.offsetHeight,
                html.clientHeight,
                html.scrollHeight,
                html.offsetHeight,
              );
              savingIdRects.current[index].limitDown = height;
            }
          }
        });
      }
    }, 500);
  }, []);

  useEventListener(
    'scroll',
    () => {
      let activated = activatedMenu;
      savingIdRects.current.forEach((item, index) => {
        if (window.scrollY >= item.limitUp && window.scrollY < item.limitDown) {
          activated = ids[index];
        }
      });
      if (activated !== activatedMenu) {
        setActiveMenu(activated);
      }
    },
    true,
  );

  const handleScrollToId = useCallback((e, id) => {
    e.preventDefault();
    window.history.pushState({}, '', `${router.pathname}#${id}`);
    scrollToElement(id);
  }, []);

  return (
    <div className={classNames(classes.root, className)}>
      <ul
        className={classNames(classes.menu, menuClassName)}
        data-test="menu-scrollable-layout-menu"
      >
        {menuConfigs.map((item) => (
          <li
            key={item.id}
            className={classes.menuItem}
            data-test="menu-scrollable-layout-menu-item"
          >
            <Link href={`${router.pathname}#${item.id}`} passHref>
              <a
                href={`${router.pathname}#${item.id}`}
                className={classNames(classes.menuLink, {
                  active: item.id === activatedMenu,
                })}
                onClick={(e) => handleScrollToId(e, item.id)}
              >
                {item.label}
              </a>
            </Link>
          </li>
        ))}
      </ul>
      <div className={classNames(classes.content, contentClassName)}>
        {children}
      </div>
    </div>
  );
};
MenuScrollableLayout.propTypes = {
  menuConfigs: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    }),
  ).isRequired,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  menuClassName: PropTypes.string,
  contentClassName: PropTypes.string,
};
MenuScrollableLayout.defaultProps = {
  className: '',
  menuClassName: '',
  contentClassName: '',
};

export default memo(MenuScrollableLayout);
