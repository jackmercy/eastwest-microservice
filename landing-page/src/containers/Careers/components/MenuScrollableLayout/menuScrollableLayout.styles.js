import { createUseStyles } from 'react-jss';

import { rem } from '../../../../utils/jss';
import COLORS from '../../../../constants/COLORS';
import BREAKPOINTS from '../../../../constants/BREAKPOINTS';

export default createUseStyles({
  root: {
    display: 'flex',
    width: '100%',
    boxSizing: 'border-box',
  },
  menu: {
    flex: '1 1 auto',
    position: 'sticky',
    top: rem(120),
    minWidth: rem(200),
    alignSelf: 'flex-start',
    padding: 0,
    margin: 0,
    listStyleType: 'none',
    borderLeft: `${rem(2)} solid ${COLORS.borderGray01}`,
  },
  menuItem: {
    color: COLORS.textBlack01,
    '& + $menuItem': { marginTop: rem(22) },
  },
  menuLink: {
    display: 'block',
    textDecoration: 'none',
    fontWeight: 600,
    fontSize: rem(12),
    lineHeight: rem(16),
    position: 'relative',
    padding: [rem(5), rem(10)],
    color: COLORS.textBlack01,
    '&::before, &::after': {
      width: rem(2),
      display: 'block',
      position: 'absolute',
      left: rem(-2),
      top: '50%',
      transform: 'translate(-35%, -50%)',
    },
    '&::before': {
      zIndex: 2,
      height: rem(26),
      borderRadius: rem(2),
      backgroundColor: COLORS.primary,
    },
    '&::after': {
      zIndex: 1,
      height: rem(34),
      backgroundColor: COLORS.bgWhite,
    },
    '&.active': {
      color: COLORS.primary,
      '&::before, &::after': { content: '""' },
    },
  },
  content: {
    width: '100%',
    maxWidth: rem(1002),
    overflow: 'hidden',
    paddingBottom: rem(160),
  },

  [`@media only screen and (max-width: ${BREAKPOINTS.maxMd}px)`]: {
    menu: { display: 'none' },
    content: { maxWidth: '100%' },
  },
  [`@media only screen and (max-width: ${BREAKPOINTS.maxSm}px)`]: {
    content: { paddingBottom: rem(100) },
  },
});
