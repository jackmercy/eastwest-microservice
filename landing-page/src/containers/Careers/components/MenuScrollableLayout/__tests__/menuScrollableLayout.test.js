import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import MenuScrollableLayout from '../MenuScrollableLayout';

describe('Testing Careers components > MenuScrollableLayout', () => {
  let wrapper;
  const props = {
    id: '123',
    title: 'CAREERS.OUR_MISSION.TITLE',
    menuConfigs: [
      { id: 'id-01', label: 'Menu-01' },
      { id: 'id-02', label: 'Menu-02' },
      { id: 'id-03', label: 'Menu-03' },
    ],
    children: <div>Content</div>,
  };

  beforeAll(() => {
    wrapper = mount(<MenuScrollableLayout {...props} />);
  });

  it('should render match snapshot', () => {
    const wrapper = shallow(<MenuScrollableLayout {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have a menu', () => {
    expect(
      wrapper.find('[data-test="menu-scrollable-layout-menu"]'),
    ).toHaveLength(1);
  });

  it('should have number of item equal with configs input', () => {
    expect(
      wrapper.find('[data-test="menu-scrollable-layout-menu-item"]'),
    ).toHaveLength(props.menuConfigs.length);
  });
});
