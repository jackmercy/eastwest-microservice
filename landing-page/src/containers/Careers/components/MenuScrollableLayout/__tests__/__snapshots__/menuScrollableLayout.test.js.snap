// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Careers components > MenuScrollableLayout should render match snapshot 1`] = `
<div
  className="root-0-2-1"
>
  <ul
    className="menu-0-2-2"
    data-test="menu-scrollable-layout-menu"
  >
    <li
      className="menuItem-0-2-3"
      data-test="menu-scrollable-layout-menu-item"
      key="id-01"
    >
      <Link
        href="some path#id-01"
        passHref={true}
      >
        <a
          className="menuLink-0-2-4 active"
          href="some path#id-01"
          onClick={[Function]}
        >
          Menu-01
        </a>
      </Link>
    </li>
    <li
      className="menuItem-0-2-3"
      data-test="menu-scrollable-layout-menu-item"
      key="id-02"
    >
      <Link
        href="some path#id-02"
        passHref={true}
      >
        <a
          className="menuLink-0-2-4"
          href="some path#id-02"
          onClick={[Function]}
        >
          Menu-02
        </a>
      </Link>
    </li>
    <li
      className="menuItem-0-2-3"
      data-test="menu-scrollable-layout-menu-item"
      key="id-03"
    >
      <Link
        href="some path#id-03"
        passHref={true}
      >
        <a
          className="menuLink-0-2-4"
          href="some path#id-03"
          onClick={[Function]}
        >
          Menu-03
        </a>
      </Link>
    </li>
  </ul>
  <div
    className="content-0-2-5"
  >
    <div>
      Content
    </div>
  </div>
</div>
`;
